modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -417,6 +436,80 @@ private void handleGetAndUpsertCommand(CommandClosure<GetAndUpsertCommand> clo)
         clo.result(new SingleRowResponse(response));
     }
 
+    /**
+     * Handler for the {@link ScanInitCommand}.
+     *
+     * @param clo Command closure.
+     */
+    private void handleScanInitCommand(CommandClosure<ScanInitCommand> clo) {
+        ScanInitCommand rangeCmd = clo.command();
+
+        IgniteUuid cursorId = rangeCmd.scanId();
+
+        Cursor<DataRow> cursor = storage.scan(key -> true);
+
+        cursors.put(
+            cursorId,
+            new CursorMeta(
+                cursor,
+                rangeCmd.requesterNodeId()
+            )
+        );
+
+        clo.result(null);
+    }
+
+    /**
+     * Handler for the {@link ScanRetrieveBatchCommand}.
+     *
+     * @param clo Command closure.
+     */
+    private void handleScanRetrieveBatchCommand(CommandClosure<ScanRetrieveBatchCommand> clo) {
+        CursorMeta cursorDesc = cursors.get(clo.command().scanId());
+
+        if (cursorDesc == null) {
+            clo.result(new NoSuchElementException("Corresponding cursor on server side not found."));
+
+            return;
+        }
+
+        List<BinaryRow> res = new ArrayList<>();
+
+        try {
+            for (int i = 0; i < clo.command().itemsToRetrieveCount() && cursorDesc.cursor().hasNext(); i++)
+                res.add(new ByteBufferRow(cursorDesc.cursor().next().valueBytes()));
+        }
+        catch (Exception e) {
+            clo.result(e);
+        }
+
+        clo.result(new MultiRowsResponse(res));
+    }
+
+    /**
+     * Handler for the {@link ScanCloseCommand}.
+     *
+     * @param clo Command closure.
+     */
+    private void handleScanCloseCommand(CommandClosure<ScanCloseCommand> clo) {
+        CursorMeta cursorDesc = cursors.get(clo.command().scanId());
+
+        if (cursorDesc == null) {
+            clo.result(null);
+
+            return;
+        }
+
+        try {
+            cursorDesc.cursor().close();
+        }
+        catch (Exception e) {
+            throw new IgniteInternalException(e);
```

Comment:
Probably not, there's nothing to be done by user with such exception.

Reply:
> The exception doesn't matter actually. But what does matter is fact that something went wrong and there is no need for user code to wait for response, because error response is the response. Otherwise, they will wait for an answer that will never come.
In case of throwing an exception from state machine raft node should be stopped and corresponding RaftException should be propagated to the client, so client won't wait for the answer any longer thаn in case of clo.result(exception),If it's not happening it's definitely a bug, however it's not in the scope of given ticket.
> BTW I don't sure about cursorDesc.cursor().close(), but cursorDesc.cursor().hasNext() throws sometimes an AssertionError, hence I propose to catch Throwable.
Well, I don't agree here. Only business logic exceptions should be processed with clo.result, all other including AssertionError should stop raft node and be propagated to client as RaftException, same as mentioned above actually.